feat: add openebs_upgrade

niladrih opened the pull request:
This PR adds the upgrade-job code for the openebs/openebs helm chart. Upgrade-job takes after the upgrade crate in openebs/mayastor-extensions.

This uses `helm upgrade --reset-then-reuse-values` and does not use yq to generate values exhaustively (unlike openebs/mayastor-extensions),

niladrih commented:
Requires PR `https://github.com/openebs/mayastor-extensions/pull/613`

niladrih committed:
feat: add openebs_upgrade

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

tiagolobocastro reviewed (changes requested):
Neither the commits nor the PR have any description whatsoever
Please highlight what this is doing, and why it's needed.

niladrih commented:
> Neither the commits nor the PR have any description whatsoever Please highlight what this is doing, and why it's needed.

Sorry! I've tried to put in a brief summary.

tiagolobocastro commented on openebs_upgrade/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+edition = "2021"
+
+[[bin]]
+name = "upgrade-job"
```
This is what goes into the container?

niladrih replied:
Yes.

tiagolobocastro commented on openebs_upgrade/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+color-eyre = "0.6.3"
+constants = { path = "../mayastor/constants" }
+futures = "0.3.31"
+k8s-openapi = { version = "0.20.0", features = ["v1_22"] }
```
```suggestion
k8s-openapi = { version = "0.22.0", features = ["v1_24"] }
```

tiagolobocastro commented on openebs_upgrade/src/bin/upgrade-job/cli_args.rs:
```diff
@@ -0,0 +1,68 @@
+pub(crate) struct CliArgs {
+    /// This is the URL for the storage REST API server.
+    #[arg(short = 'e', long)]
+    pub(crate) rest_endpoint: Url,
```
pass ca-certificate like what's being done on mayastor now?

tiagolobocastro commented on openebs_upgrade/src/constants.rs:
```diff
@@ -0,0 +1,7 @@
+pub use constants::{DS_CONTROLLER_REVISION_HASH_LABEL_KEY, UMBRELLA_CHART_NAME};
+use semver::Version;
+pub use upgrade::common::constants::IO_ENGINE_LABEL;
+pub const UMBRELLA_CHART_VERSION_LOWERBOUND: Version = Version::new(3, 0, 0);
```
doc comments for the constants here

tiagolobocastro commented on openebs_upgrade/src/helm/data_plane_upgrader.rs:
```diff
@@ -0,0 +1,27 @@
+}
+
+impl DataPlaneUpgrader {
+    pub async fn run(
```
document that this does

tiagolobocastro commented on openebs_upgrade/src/helm/metadata.rs:
```diff
@@ -0,0 +1,43 @@
+use tracing::debug;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct HelmChartMetadata {
```
doc comment

tiagolobocastro commented on openebs_upgrade/src/helm/upgrade.rs:
```diff
@@ -0,0 +1,161 @@
+pub mod config;
+pub mod upgrader;
+
+pub async fn upgrade(
```
doc comment

tiagolobocastro commented on openebs_upgrade/src/helm/upgrade.rs:
```diff
@@ -0,0 +1,161 @@
+    Ok(())
+}
+
+async fn mayastor_data_plane_upgrade(
```
Can't we make use of the mayastor crate to do this?

tiagolobocastro commented on openebs_upgrade/src/utils.rs:
```diff
@@ -0,0 +1,55 @@
+    Ok(output)
+}
+
+pub(crate) async fn check_path(
```
doc comment

tiagolobocastro commented on openebs_upgrade/src/utils.rs:
```diff
@@ -0,0 +1,55 @@
+        })
+}
+
+pub(crate) async fn flatten_task<T>(handle: JoinHandle<Result<T>>) -> Result<T> {
```
Seems like it's just flattening but it's waiting for it?

```suggestion
pub(crate) async fn join_task<T>(handle: JoinHandle<Result<T>>) -> Result<T> {
```

niladrih replied:
It's flattening the output of a task, it'll wait for the task to finish, yes.

niladrih replied:
Going with `joined_flatten()`...

niladrih committed:
feat(upgrade-job): install CRDs intelligently
When upgrading from OpenEBS v3.x to v4.y, if the labels for the yaml
keys for release name, etc. have a different value, the installation
fails. This commit checks for CRDs and if the ones from the openebs/
openebs chart or its openebs flavoured sub-charts exist, then their
installation is skipped by means of helm chart values toggles.

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

niladrih committed:
chore(deps): update mayastor-extensions git submodule

Signed-off-by: Niladri Halder <niladri.halder26@gmail.com>

tiagolobocastro commented on Cargo.toml:
```diff
@@ -1,6 +1,7 @@
 [workspace]
 members = [
     "plugin",
+    "openebs_upgrade"
```
```suggestion
    "openebs-upgrade"
```

tiagolobocastro commented on openebs_upgrade/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+[package]
+name = "openebs_upgrade"
```
```suggestion
name = "openebs-upgrade"
```

niladrih replied:
Done.

niladrih commented:
bors merge
